### PR TITLE
[Fix #2968] Added DocumentationMethod Cop.

### LINKED ISSUE #2968: check the comments or document before each function?
## Expected behavior

Hi,
i find the rule 'Style/Documentation' can check documentation before class or module 
i want to add a rule to check the comment or document before each function.
i do not find this kind of rule?
can rubocop support it ?

## RuboCop version

Include the output of `rubocop -V`:

```
$ rubocop -V
0.35.1( running on ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-linux])
```

### PR DESCRIPTION
**Pull request by sooyang**
This cop checks for missing top-level documentation of methods as suggested in issue #2968.
The default is set to false due to pre-existing methods that do not contain the top-level documentation. 
Private and Protected methods will not be scanned.
-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: lib/rubocop/cop/style/documentation_method.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for missing top-level documentation of
+      # methods. Private methods are excluded from this cop
+      class DocumentationMethod < Cop
+        include AnnotationComment
+
+        MSG = 'Missing top-level %s documentation method comment.'.freeze
```
Comment (jonas054) [inline lib/rubocop/cop/style/documentation_method.rb]: I don't quite get this message. I'd go for "Missing documentation comment for public method.".

File: spec/rubocop/cop/style/documentation_method_spec.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                          'Keywords' => %w(TODO FIXME OPTIMIZE HACK REVIEW)
+                        },
+                        'Style/DocumentationMethod' => {
+                          'Enabled' => true
+                        })
+  end
+
+  it 'registers an offense for non-empty method' do
+    inspect_source(cop,
+                   ['class My_Class',
```
Comment (jonas054) [inline spec/rubocop/cop/style/documentation_method_spec.rb]: `My_Class` is a bit non-standard. Call it `MyClass`.

File: config/enabled.yml [hunk 3]
```diff
@@ -176,6 +176,13 @@ Style/PreferredHashMethods:
   StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#hash-key'
   Enabled: true
 
+Style/DocumentationMethod:
```
Comment (jonas054) [inline config/enabled.yml]: The configuration for this cop should be in the file `disabled.yml`.

File: spec/rubocop/cop/style/documentation_method_spec.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# encoding: utf-8
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe RuboCop::Cop::Style::DocumentationMethod do
```
Comment (jonas054) [inline spec/rubocop/cop/style/documentation_method_spec.rb]: You need more examples. Methods in modules, methods on the top level (outside a class), different ways of declaring a method private (`private def method` and `private :method`), combinations of public and private methods.

File: CHANGELOG.md [hunk 5]
```diff
@@ -56,6 +56,7 @@
 * [#3127](https://github.com/bbatsov/rubocop/pull/3127): New cop `Lint/InheritException` checks for error classes inheriting from `Exception`, and instead suggests `RuntimeError` or `StandardError`. ([@drenmi][])
 * Add new `Performance/PushSplat` cop. ([@segiddins][])
 * [#3089](https://github.com/bbatsov/rubocop/issues/3089): Add new `Rails/Exit` cop. ([@sgringwe][])
+* [#2968](https://github.com/bbatsov/rubocop/issues/2968): Add new `Style/DocumentationMethod` cop. ([@sooyang][])
```
Comment (jonas054) [inline CHANGELOG.md]: Move this line to the `master (unreleased)` section.

**Comment by bbatsov** [comment]
@sooyang ping :-)

**Comment by sooyang** [comment]
@bbatsov @jonas054 
noted :)

**Comment by sooyang** [comment]
@bbatsov @jonas054 
updated :)

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,7 @@
 # Change log
 
 ## master (unreleased)
+* [#2968](https://github.com/bbatsov/rubocop/issues/2968): Add new `Style/DocumentationMethod` cop. ([@sooyang][])
```
Comment (bbatsov) [inline CHANGELOG.md]: This should be under `New features`.

File: lib/rubocop/cop/style/documentation_method.rb [hunk 7]
```diff
@@ -0,0 +1,89 @@
+# encoding: utf-8
+# frozen_string_literal: true
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for missing documentation comment for public method.
```
Comment (bbatsov) [inline lib/rubocop/cop/style/documentation_method.rb]: Add some examples here of code that would be flagged by the cop. You can see other cops for references - e.g. https://github.com/Drenmi/rubocop/blob/635403972a3b768d9d709f04eed52090d7738d63/lib/rubocop/cop/style/numeric_predicate.rb

File: lib/rubocop/cop/style/documentation_method.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+        def_node_matcher :constant_definition?, '{def casgn}'
+
+        def on_def(node)
+          _name, body = *node
+
+          line = node.loc.line
+          return if node.ancestors.first.to_a.include? method_type
+          return if (processed_source[0..line].map(&:strip) &
+            %w(private protected)).any?
```
Comment (bbatsov) [inline lib/rubocop/cop/style/documentation_method.rb]: I'd extract this into a constant.

File: lib/rubocop/cop/style/documentation_method.rb [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+          return if namespace?(body)
+          ast_with_comments = processed_source.ast_with_comments
+          return if associated_comment?(node, ast_with_comments)
+          add_offense(node, :keyword, format(MSG, :module))
+        end
+
+        def on_method_def(node, _method_name, _args, _body)
+          _name, _body = *node
```
Comment (bbatsov) [inline lib/rubocop/cop/style/documentation_method.rb]: This method is pretty much the same as the one above. You should extract the shared code somewhere.

File: config/disabled.yml [hunk 10]
```diff
@@ -23,6 +23,13 @@ Style/Copyright:
   Description: 'Include a copyright notice in each file before any code.'
   Enabled: false
 
+Style/DocumentationMethod:
+  Description: 'Document methods.'
```
Comment (bbatsov) [inline config/disabled.yml]: public methods

**Comment by sooyang** [comment]
@bbatsov added

File: CHANGELOG.md [hunk 11]
```diff
@@ -10,6 +10,7 @@
 * Add new `Style/TernaryParentheses` cop. ([@drenmi][])
 * [#3136](https://github.com/bbatsov/rubocop/issues/3136): Add config for `UselessAccessModifier` so it can be made aware of ActiveSupport's `concerning` and `class_methods` methods. ([@maxjacobson][])
 * [#3128](https://github.com/bbatsov/rubocop/issues/3128): Add new `Rails/SaveBang` cop. ([@QuinnHarris][])
+* [#2968](https://github.com/bbatsov/rubocop/issues/2968): Add new `Style/DocumentationMethod` cop. ([@sooyang][])
```
Comment (bbatsov) [inline CHANGELOG.md]: You'll have to move this to the `master` branch. We just release 0.42.

**Comment by sooyang** [comment]
@bbatsov  moved

**Pull request merged**